### PR TITLE
Face recognition POC with Facenet model using TF

### PR DESCRIPTION
**Pull request by rahul003**
I put together a prototype for face recognition using the coordinates of faces already detected from Pigo. Ref: https://github.com/photoprism/photoprism/issues/22#issuecomment-871576686 

FaceNet seems to be the best open source pretrained model available for face recognition. I got the pretrained model checkpoint from [FaceNet's original repo](https://github.com/davidsandberg/facenet) and created a SavedModel. I can share the script used to create SavedModel if needed. I added this savedModel file to this PR. (The model used is  [20180402-114759](https://github.com/davidsandberg/facenet#pre-trained-models) ). Since PhotoPrism already links against TF, I thought this was the best way to proceed with face recognition. It seems to work quite well too on the test images in `face` folder. 

Flow here is the following
- Crop image using the coordinates and size of face returned from Pigo detector, to create a jpeg for each face
- Resize image to 160x160 as needed for the model
- Run inference using Facenet with this face image, which returns a (512,) embedding vector.
- Compute euclidian distance using these embedding vectors which gives us a score of similarity (lower the distance, more similar). Threshold needs to be adjusted based on real world data and the UX plan in Photoprism for face recognition. In the unit test I set a value for which most pairs were identified correctly (51/55 possible pairs of faces in images)
 
Let me know your thoughts on using this model and this approach for Face recognition. To proceed there needs to be some thought on the plan for comparing faces across the whole library of the user. There's of course normalization and other processing (whitening is one step done in Facenet repo which I haven't done yet) which could improve accuracy. Switching to an efficient implementation of Euclidian distance, vectorization etc can improve performance. 

- [x] P.S. CI is failing because the order of processing files seems different, I'll rework the test.

### CODE REVIEW

**Comment by lastzero** [comment]
Thank you very much! This sounds perfect... still busy with other tasks, doing my best to test your PR as soon as possible :+1:

**Comment by lastzero** [comment]
Finally have a bit of time, after upgrading packages and improving our multi-arch Docker image earlier this week...

Guess we can merge this, but without `assets/facenet/saved_model.pb`. Should put that on `dl.photoprism.org` to reduce the git repo size - especially in case the model needs to be updated.

**Comment by lastzero** [comment]
@rahul003 Face embeddings will now be stored in the markers table while indexing. To improve performance, it probably makes sense to cache face thumbs so that they don't need to be recreated every time. As your code comment suggests, it probably also makes sense to normalize them further for best results (whiten?). Currently reading the FaceNet paper to get a better understanding.

Are there any conditions under which FaceNet may return more than one embedding per face (file)? I've noticed the return value is a nested array.

**Comment by lastzero** [comment]
From my understanding, it seems best to perform clustering & matching in a background worker after indexing, ideally when no other workers are active. Results can be stored using `Face` and `Person` entities, so that we can match different looking faces with the same person if needed.

I'd try to cluster embeddings using https://github.com/mpraski/clusters, unless someone knows a better library / approach.

**Comment by rahul003** [comment]
> @rahul003 Face embeddings will now be stored in the markers table while indexing. To improve performance, it probably makes sense to cache face thumbs so that they don't need to be recreated every time. As your code comment suggests, it probably also makes sense to normalize them further for best results (whiten?). Currently reading the FaceNet paper to get a better understanding.

Nice! Thanks for following up on this. I see that changes from this are merged, so closing this PR.

 
> Are there any conditions under which FaceNet may return more than one embedding per face (file)? I've noticed the return value is a nested array.

I believe that it'll return more than one embedding if we send a batch of faces as opposed to one face.

**Comment by rahul003** [comment]
> From my understanding, it seems best to perform clustering & matching in a background worker after indexing, ideally when no other workers are active. Results can be stored using `Face` and `Person` entities, so that we can match different looking faces with the same person if needed.
> 
> I'd try to cluster embeddings using https://github.com/mpraski/clusters, unless someone knows a better library / approach.

I agree. It can be intensive I think, best to separate our from initial indexing. I notice that Apple Photos also does it in the background later.

**Comment by janstadt** [comment]
Hey this looks great! Im a bit late to the party but will this/photoprism (or facenet by proxy) support the coral usb accelerator for helping out the cpu for detection? That would be a great addition to this of the detection process is intensive.

**Comment by lastzero** [comment]
It's possible, actually just other TensorFlow libs required. Nothing we need to implement in photoprism directly.

**Comment by MasterGoofy** [comment]
Do you know when this feature will be released?

**Comment by lastzero** [comment]
PhotoPrism doesn't require changes. You "just" need to install a version of the TensorFlow library with support for your hardware. We currently don't have time to compile a custom version of TensorFlow for everyone, so it's only for experienced users so to say. On the other hand, we also didn't reach our funding goal yet. Once that is the case, it's easier for us to help with installation and configuration. It's not that we don't want to.

**Comment by Helvio88** [comment]
What is the status of this? Is there a way to hack-implement this? I have an idle Google Coral and my CPU is running 90%+ lol

**Comment by graciousgrey** [comment]
>  Is there a way to hack-implement this?

You "just" need to install a version of the TensorFlow library with support for your hardware.
